keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,585 @@
+- one per Pod lifecycle to remove finalizers
+- one for each Job sync, to track the intermediate state.
+
+Note that there no new calls in the Pod creation path.
+
+On the other hand, to update the Job status once a Pod finishes, we need a total
+2 new API calls compared to the legacy algorithm. If more than one Pod finish at
+a given time, we add `n + 1` API calls, where `n` is the number finished Pods.
+
+Consider a Job with multiple Pods. With a 50 QPS limit in the job controller
+client, the controller should be able to process between 2000 to 3000 Pods.
+
+The increase in API calls is justified for the following reasons:
+- The legacy tracking cannot handle a big number of terminated Pods, across
+  any number of Jobs, at a given point.
+- In the entirety of its lifecycle, a Pod requires at least 8 API calls,
+  including status updates and events.
+  
+However, in order to prevent Jobs with big number of Pods from starving Jobs
+with fewer Pods, the Job controller might skip status updates until enough
+Pods have accumulated or enough time has passed. See [Algorithm](#algorithm)
+below for more details.
+   
+#### Bigger Job status
+
+Job status can temporarily grow if too many Pods finish at the same time or
+if the Job controller is down for some time. In this case, we do partial
+status updates. See [Algorithm](#algorithm) below.
+
+#### Unprotected Job status endpoint
+
+Changes in the status not produced by the Job controller in
+kube-controller-manager could affect the Job tracking. Cluster administrators
+should make sure to protect the Job status endpoint via RBAC.
+
+## Design Details
+
+### API changes
+
+The Job status gets a new struct to hold the uncounted Pods before they are
+added to the counters.
+
+```golang
+type JobStatus struct {
+    Succeeded int32
+    Failed    int32
+    ...
+
+    // UncountedTerminatedPods holds UIDs of Pods that have finished but
+    // haven't been accounted in the counters.
+    // If nil, Job tracking doesn't make use of this structure.
+    // +optional
+    UncountedTerminatedPods *UncountedTerminatedPods
+}
+
+// UncountedTerminatedPods holds UIDs of Pods that have finished but haven't
+// been accounted in Job status counters.
+type UncountedTerminatedPods struct {
+    // Succeeded holds UIDs of succeeded Pods.
+    Succeeded []types.UID
+    // Succeeded holds UIDs of failed Pods.
+    Failed    []types.UID
+}
+```
+
+Note: the final name of the field `uncountedTerminatedPods` will be decided
+during API review.
+
+### Algorithm
+
+The following algorithm updates the status counters without relying on finished
+Pods to be present indefinitely. The algorithm assumes that the Job controller
+could be stopped at any point and executed again from the first step without
+losing information. Generally, all the steps happen in a single Job sync
+cycle.
+
+1. The Job controller calculates the number of succeeded Pods as the sum of:
+   - `.status.succeeded`,
```

Comment:
@lavalamp?

Reply:
@janetkuo I don't think we are changing this significantly much, there are slight variances how this will get calculated during execution, but the final result will be the same. Nobody should ever rely on how this number is being updated, nor partial results. I'm not personally worried about that breaking API compatibility.

Reply:
SGTM